File: apps/web/src/api/session_notes.ts
import { apiClient } from './client'

export type SessionParticipant = 'parent' | 'patient'

export interface SessionNote {
  id: string
  patient_id: string
  organization_id: string
  practitioner_id: string
  session_type: string | null
  participants: SessionParticipant[]
  tags: string[]
  session_date: string
  content: string
  /** Written by Float from a recorded session, until the clinician approves it. */
  is_draft?: boolean
  source?: 'typed' | 'recording'
  /** Who said what, in order. Speaker keys like "1:2". */
  transcript?: { speaker: string; text: string }[] | null
  speaker_names?: Record<string, string> | null
  created_at: string
  updated_at: string
}

export interface CreateSessionNote {
  participants: SessionParticipant[]
  tags: string[]
  session_date?: string
  content: string
}

export interface UpdateSessionNote {
  participants?: SessionParticipant[]
  tags?: string[]
  session_date?: string
  content?: string
  speaker_names?: Record<string, string>
  /** false approves a draft. */
  is_draft?: false
}

export const getSessionNotes = async (patientId: string): Promise<SessionNote[]> => {
  const response = await apiClient.get(`/patients/${patientId}/notes`)
  return response.data
}

export const createSessionNote = async (patientId: string, data: CreateSessionNote): Promise<SessionNote> => {
  const response = await apiClient.post(`/patients/${patientId}/notes`, data)
  return response.data
}

export const updateSessionNote = async (noteId: string, data: UpdateSessionNote): Promise<SessionNote> => {
  const response = await apiClient.put(`/notes/${noteId}`, data)
  return response.data
}

export const deleteSessionNote = async (noteId: string): Promise<void> => {
  await apiClient.delete(`/notes/${noteId}`)
}

File: backend/app/api/routers/session_notes.py
import uuid
from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import datetime

from app.core.database import get_db
from app.models.session_note import SessionNote
from app.models.user import User
from app.models.patient import PractitionerProfile
from app.api.routers.patients import get_practitioner_context


# --- Schemas ---

class SessionNoteCreate(BaseModel):
    session_type: str
    session_date: Optional[date] = None
    content: str


class SessionNoteUpdate(BaseModel):
    session_type: Optional[str] = None
    session_date: Optional[date] = None
    content: Optional[str] = None


class SessionNoteResponse(BaseModel):
    id: uuid.UUID
    patient_id: uuid.UUID
    organization_id: uuid.UUID
    practitioner_id: uuid.UUID
    session_type: str
    session_date: date
    content: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# --- Router ---

router = APIRouter(tags=["session-notes"])


@router.get(
    "/patients/{patient_id}/notes",
    response_model=list[SessionNoteResponse]
)
async def list_session_notes(
    patient_id: uuid.UUID,
    context: tuple = Depends(get_practitioner_context),
    db: AsyncSession = Depends(get_db)
):
    _, practitioner = context
    result = await db.execute(
        select(SessionNote)
        .where(
            SessionNote.patient_id == patient_id,
            SessionNote.organization_id == practitioner.organization_id
        )
        .order_by(SessionNote.session_date.desc(), SessionNote.created_at.desc())
    )
    return result.scalars().all()


@router.post(
    "/patients/{patient_id}/notes",
    response_model=SessionNoteResponse,
    status_code=status.HTTP_201_CREATED
)
async def create_session_note(
    patient_id: uuid.UUID,
    data: SessionNoteCreate,
    context: tuple = Depends(get_practitioner_context),
    db: AsyncSession = Depends(get_db)
):
    _, practitioner = context
    note = SessionNote(
        patient_id=patient_id,
        organization_id=practitioner.organization_id,
        practitioner_id=practitioner.id,
        session_type=data.session_type,
        session_date=data.session_date or date.today(),
        content=data.content
    )
    db.add(note)
    await db.commit()
    await db.refresh(note)
    return note


@router.put(
    "/notes/{note_id}",
    response_model=SessionNoteResponse
)
async def update_session_note(
    note_id: uuid.UUID,
    data: SessionNoteUpdate,
    context: tuple = Depends(get_practitioner_context),
    db: AsyncSession = Depends(get_db)
):
    _, practitioner = context
    result = await db.execute(
        select(SessionNote).where(
            SessionNote.id == note_id,
            SessionNote.organization_id == practitioner.organization_id
        )
    )
    note = result.scalar_one_or_none()
    if not note:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session note not found"
        )

    if data.session_type is not None:
        note.session_type = data.session_type
    if data.session_date is not None:
        note.session_date = data.session_date
    if data.content is not None:
        note.content = data.content

    await db.commit()
    await db.refresh(note)
    return note


@router.delete(
    "/notes/{note_id}",
    status_code=status.HTTP_204_NO_CONTENT
)
async def delete_session_note(
    note_id: uuid.UUID,
    context: tuple = Depends(get_practitioner_context),
    db: AsyncSession = Depends(get_db)
):
    _, practitioner = context
    result = await db.execute(
        select(SessionNote).where(
            SessionNote.id == note_id,
            SessionNote.organization_id == practitioner.organization_id
        )
    )
    note = result.scalar_one_or_none()
    if not note:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session note not found"
        )

    await db.delete(note)
    await db.commit()
